nixos/modules/services/networking/lokinet.nix

```
@@ -0,0 +1,144 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.lokinet;
+  settingsFormat = pkgs.formats.ini { listsAsDuplicateKeys = true; };
+  configFile = settingsFormat.generate "lokinet.ini" (lib.filterAttrsRecursive (n: v: v != null) cfg.settings);
+in with lib; {
+  options.services.lokinet = {
+    enable = mkEnableOption "Lokinet daemon.";
+
+    dataDir = mkOption {
+      type = types.str;
+      default = "/var/lib/lokinet";
+      description = "The directory where Lokinet stores its data files.";
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.lokinet;
+      defaultText = literalExpression "pkgs.lokinet";
+      description = "Lokinet package to use.";
+    };
+
+    useLocally = mkOption {
+      type = types.bool;
+      default = false;
+      example = true;
+      description = "Whether to use Lokinet locally.";
+    };
+
+    settings = mkOption {
+      type = with types;
+        submodule {
+          freeformType = settingsFormat.type;
+
+          options = {
+            dns = {
+              bind = mkOption {
+                type = str;
+                default = "127.3.2.1";
+                example = "127.3.2.1";
+                description = "Address to bind to for handling DNS requests.";
+              };
+
+              upstream = mkOption {
+                type = listOf str;
+                default = [ "1.1.1.1" ];
```

```suggestion
                default = [ ];
```
Or is that essential for functionality? Why can't we rely on the DNS the user has set up?

---

I will just change that to 9.9.9.10 since it's now the default with 0.9.9.

---

Thats still not that nice. Software phoning DNS servers itself instead of relying what is configured on the system is annoying. Why can't we use what is written into /etc/resolve.conf?

---

> Thats still not that nice. Software phoning DNS servers itself instead of relying what is configured on the system is annoying. Why can't we use what is written into /etc/resolve.conf?

because parsing /etc/resolv.conf (NOT /etc/resolve.conf) is abysmal, patches welcome, it is on our todo list, so much other shit is on this list that got bumped up higher priority (like the macos port that i dont want to exist).

---

additionally there are like 5 distinct slightly conflicting ways to set dns on linux, then you network manager mangles all the settings because it thinks it knows better.

---

You are not supposed to parse resolv.conf yourself. There are libraries for that like glibs resolver. Also most programming language should just do that on their own and provide you a way to resolve domain names.